tests/src/main/java/net/neoforged/neoforge/debug/data/CustomFeatureFlagsTests.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.debug.data;
+
+import net.minecraft.network.chat.Component;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.server.packs.PackType;
+import net.minecraft.server.packs.repository.Pack;
+import net.minecraft.server.packs.repository.PackSource;
+import net.minecraft.world.flag.FeatureFlag;
+import net.minecraft.world.flag.FeatureFlagSet;
+import net.minecraft.world.flag.FeatureFlags;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.level.Level;
+import net.neoforged.neoforge.event.AddPackFindersEvent;
+import net.neoforged.neoforge.event.server.ServerStartedEvent;
+import net.neoforged.neoforge.registries.DeferredItem;
+import net.neoforged.testframework.DynamicTest;
+import net.neoforged.testframework.annotation.ForEachTest;
+import net.neoforged.testframework.annotation.TestHolder;
+
+@ForEachTest(groups = "data.feature_flags")
+public class CustomFeatureFlagsTests {
+    @TestHolder(description = "Tests that feature flag packs get shown in the experiments screen", enabledByDefault = true)
+    static void testFeatureFlagPacks(final DynamicTest test) {
+        test.framework().modEventBus().addListener((AddPackFindersEvent event) -> {
+            event.addPackFinders(
+                    ResourceLocation.fromNamespaceAndPath("neotests", "feature_flag_test_packs/flag_test_pack"),
+                    PackType.SERVER_DATA,
+                    Component.literal("Custom FeatureFlag test pack"),
+                    PackSource.FEATURE,
+                    false,
+                    Pack.Position.TOP);
+
+            // Add 6 additional packs to visually overflow the vanilla experiments screen
+            for (int i = 0; i < 6; i++) {
+                event.addPackFinders(
+                        ResourceLocation.fromNamespaceAndPath("neotests", "feature_flag_test_packs/flag_test_pack_" + i),
+                        PackType.SERVER_DATA,
+                        Component.literal("Custom FeatureFlag test pack " + i),
+                        PackSource.FEATURE,
+                        false,
+                        Pack.Position.TOP);
+            }
+
+            test.pass();
+        });
+    }
+
+    @TestHolder(description = "Verifies that registered objects using a custom feature flag are not accessible without the feature flag being enabled", enabledByDefault = true)
+    static void testFeatureGating(final DynamicTest test) {
+        test.framework().modEventBus().addListener((AddPackFindersEvent event) -> event.addPackFinders(
+                ResourceLocation.fromNamespaceAndPath("neotests", "feature_flag_test_packs/gating_test_pack"),
+                PackType.SERVER_DATA,
+                Component.literal("Custom FeatureFlag gating test pack"),
+                PackSource.FEATURE,
+                true,
```

Comment:
Shouldn't this be `false`? It being `true` means this pack is always active and can not be disabled, even by the user in the experiments screen.
Pack should be able to be disabled to validate that the flagged items do indeed get disabled when the flags are disabled.

Other than this everything looks good to me.

Reply:
That's fully intentional in this particular case, otherwise the test wouldn't work at all on the game test server since nothing would enable the pack and it would require intervention to make the test work in singleplayer, both of which defeat the point of the test being automated.

Reply:
Suppose you do make a point here but I would have had the pack be enabled via `server.properties`->`initial-enabled-packs` (if possible for `GameTestServer`s) or similar means rather than forcing the pack to always be enabled.

Reply:
A mod has no reasonable way to modify that and it also wouldn't help in singleplayer.